### PR TITLE
schedule: transfer leader in batch when do eviction

### PR DESCRIPTION
**Pull request by shafreeck**
Signed-off-by: Shafreeck Sea <shafreeck@gmail.com>

### What problem does this PR solve? 

This PR tries to improve the performance of leader transferring by returning multiple operators in batch.

* Move the condition check of max waiting operators into `checkAddOperator`
* Support to add operators partially when the number of operators returned by `Schedule` exceeds the limitation of max waiting operators

### Check List 

Tests 

 - Unit test
 - Integration test

### CODE REVIEW

File: server/schedulers/evict_leader.go [hunk 1]
```diff
@@ -242,6 +242,42 @@ func (s *evictLeaderScheduler) Schedule(cluster opt.Cluster) []*operator.Operato
 		op.Counters = append(op.Counters, schedulerCounter.WithLabelValues(s.GetName(), "new-operator"))
 		ops = append(ops, op)
 	}
+	return ops
+}
+
+func (s *evictLeaderScheduler) uniqeAppend(dst []*operator.Operator, src ...*operator.Operator) []*operator.Operator {
```
Comment (lhy1024) [inline server/schedulers/evict_leader.go]:
```suggestion
func (s *evictLeaderScheduler) uniqueAppend(dst []*operator.Operator, src ...*operator.Operator) []*operator.Operator {
```

**Comment by shafreeck** [comment]
@nolouch Will do some test later

File: server/schedulers/evict_leader.go [hunk 2]
```diff
@@ -36,8 +36,11 @@ const (
 	// EvictLeaderName is evict leader scheduler name.
 	EvictLeaderName = "evict-leader-scheduler"
 	// EvictLeaderType is evict leader scheduler type.
-	EvictLeaderType     = "evict-leader"
-	lastStoreDeleteInfo = "The last store has been deleted"
+	EvictLeaderType = "evict-leader"
+	// EvictLeaderBatchSize is the number of operators to to transfer
+	// leaders by one scheduling
+	EvictLeaderBatchSize = 3
```
Comment (rleungx) [inline server/schedulers/evict_leader.go]: How about making it configurable?

Comment (shafreeck) [reply server/schedulers/evict_leader.go]: Leader transferring is a very light operation with a tiny impact on the performance, there is little necessity for a user to customize it.

**Comment by shafreeck** [comment]
After this PR, in the test environment, PD transferred 6156 leaders in 21s, indicates 293/s, which gains 3 times performance improvement.

File: server/schedule/operator_controller.go [hunk 3]
```diff
@@ -264,15 +265,18 @@ func (oc *OperatorController) AddWaitingOperator(ops ...*operator.Operator) bool
...
 	}
 	oc.wop.PutOperator(op)
 	operatorWaitCounter.WithLabelValues(op.Desc(), "put").Inc()
-	// This step is especially for the merge operation.
+	oc.wopStatus.ops[desc]++
 	if len(ops) > 1 {
-		oc.wop.PutOperator(ops[1])
+		for _, op := range ops[1:] {
+			oc.wop.PutOperator(op)
+			oc.wopStatus.ops[desc]++
```
Comment (nolouch) [inline server/schedule/operator_controller.go]: After the change, the odd number may let the merge operator only push one operator to the waiting operator and the other one exceed the max limit. please add a unit test to make sure.

Comment (shafreeck) [reply server/schedule/operator_controller.go]: The check `oc.wopStatus.ops[desc] >= oc.cluster.GetSchedulerMaxWaitingOperator()` has been done above.

File: server/config/config.go [hunk 4]
```diff
@@ -641,7 +641,7 @@ const (
 	// defaultHotRegionCacheHitsThreshold is the low hit number threshold of the
 	// hot region.
 	defaultHotRegionCacheHitsThreshold = 3
-	defaultSchedulerMaxWaitingOperator = 3
+	defaultSchedulerMaxWaitingOperator = 5
```
Comment (nolouch) [inline server/config/config.go]: If the user upgrades from 3.0 to 4.0, the value is still 3.

File: server/schedule/operator_controller.go [hunk 5]
```diff
@@ -316,11 +320,13 @@ func (oc *OperatorController) PromoteWaitingOperator() {
 				operatorWaitCounter.WithLabelValues(op.Desc(), "promote_canceled").Inc()
 				_ = op.Cancel()
 				oc.buryOperator(op)
+				oc.wopStatus.ops[op.Desc()]--
```
Comment (rleungx) [inline server/schedule/operator_controller.go]: For merge, `GetOperator` will return two operators, but `oc.wopStatus.ops[op.Desc()]--` should only happen once.

File: server/schedule/operator_controller.go [hunk 6]
```diff
@@ -252,6 +252,7 @@ func (oc *OperatorController) AddWaitingOperator(ops ...*operator.Operator) bool
 			oc.buryOperator(op)
 		}
 		oc.Unlock()
+		log.Debug("checkAddOperator return false")
```
Comment (rleungx) [inline server/schedule/operator_controller.go]: prefer to make these debug messages more readable if we decide to keep them

**Comment by shafreeck** [comment]
Refactor the logic of adding waiting operators

* Move the condition check of max waiting operators into `checkAddOperator`
* Support to add operators partially when the number of operators returned by Schedule exceeds the limitation of max waiting operators

PTAL @nolouch @rleungx @lhy1024

File: server/schedule/operator_controller_test.go [hunk 7]
```diff
@@ -637,3 +637,27 @@ func checkRemoveOperatorSuccess(c *C, oc *OperatorController, op *operator.Opera
...
+	var batch []*operator.Operator
+	for i := uint64(0); i < cluster.GetSchedulerMaxWaitingOperator()+1; i++ {
+		start := fmt.Sprintf("%da", i)
+		end := fmt.Sprintf("%db", i)
+		region := newRegionInfo(i, start, end, 1, 1, []uint64{101, 1}, []uint64{101, 1})
+		cluster.PutRegion(region)
+		peer := &metapb.Peer{
+			StoreId: 2,
+		}
+		op, err := operator.CreateAddPeerOperator("add-peer", cluster, region, peer, operator.OpBalance)
```
Comment (nolouch) [inline server/schedule/operator_controller_test.go]: Can you add a case which contains merge operaotr? it's special.

File: server/schedule/operator_controller.go [hunk 8]
```diff
@@ -242,40 +242,51 @@ func (oc *OperatorController) PushOperators() {
...
-		oc.Unlock()
-		return false
-	}
-
-	op := ops[0]
-	desc := op.Desc()
-	if oc.wopStatus.ops[desc] >= oc.cluster.GetSchedulerMaxWaitingOperator() {
-		operatorWaitCounter.WithLabelValues(op.Desc(), "exceed_max").Inc()
-		for _, op := range ops {
+		if !oc.checkAddOperator(op) {
```
[the remark this answers is not in the document]
Comment (shafreeck) [reply server/schedule/operator_controller.go]: No, it has variable parameters

**Comment by rleungx** [comment]
CI failed.

**Pull request merged**